net/inet/ipv4_setsockopt.c

```diff
@@ -184,16 +189,41 @@ int ipv4_setsockopt(FAR struct socket *psock, int option,
         }
         break;
 
-      /* The following IPv4 socket options are defined, but not implemented */
-
-      case IP_MULTICAST_IF:           /* Set local device for a multicast
-                                       * socket */
       case IP_MULTICAST_TTL:          /* Set/read the time-to-live value of
                                        * outgoing multicast packets */
+        {
+          if (psock->s_type != SOCK_DGRAM ||
+              value_len != sizeof(int))
```

Comment:
This is not correct.  I should be able to use `unsigned char ttl;` as the input to set the ttl.

Reply:
bsd and as well as tldp and ltp all use uchar
```
     The IP_MULTICAST_TTL option changes the time-to-live (TTL)	for outgoing
     multicast datagrams in order to control the scope of the multicasts:

     u_char ttl;     /*	range: 0 to 255, default = 1 */
     setsockopt(s, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, sizeof(ttl));
```
The linux man 7 ip page is unfortunate because it seems to be playing loose with the word `integer`
```
       IP_MULTICAST_TTL (since Linux 1.2)
              Set or read the time-to-live value of outgoing multicast pack‐
              ets for this socket.  It is very important for multicast pack‐
              ets to set the smallest TTL possible.  The default is 1 which
              means that multicast packets don't leave the local network un‐
              less the user program explicitly requests it.  Argument is an
              integer.
```
However since there own test suite uses `uchar` and every usage of it I have seen does as well I think it is clear.
https://github.com/linux-test-project/ltp/blob/master/testcases/network/multicast/mc_opts/mc_verify_opts.c

Reply:
For Linux for most of these options both signed int and unsigned char are supported and they just dereference and cast the value to in in the case of unsigned char to a signed int.

Reply:
I applied this change and the ttl of 20 that I requested is set (instead of the default 64).  I also verified the signed int with a different value.
```c
diff --git a/net/inet/ipv4_setsockopt.c b/net/inet/ipv4_setsockopt.c
index 995b750b11..22e19f7860 100644
--- a/net/inet/ipv4_setsockopt.c
+++ b/net/inet/ipv4_setsockopt.c
@@ -193,14 +193,23 @@ int ipv4_setsockopt(FAR struct socket *psock, int option,
                                        * outgoing multicast packets */
         {
           if (psock->s_type != SOCK_DGRAM ||
-              value_len != sizeof(int))
+              !(value_len == sizeof(int) ||
+                value_len == sizeof(uint8_t)))
             {
               ret = -EINVAL;
             }
           else
             {
               FAR struct udp_conn_s *conn;
-              int ttl = *(FAR int *)value;
+              int ttl;
+              if (value_len == sizeof(int))
+                {
+                  ttl = *(FAR int *)value;
+                }
+              else
+                {
+                  ttl = (int)*(FAR uint8_t *)value;
+                }
 
               if (ttl <= 0 || ttl > 255)
                 {
```
![image](https://user-images.githubusercontent.com/173245/101238742-fb0f2480-3696-11eb-8f9d-d44545c84bdb.png)

Reply:
Hi @btashton 

You are right. I just checked the linux source code and found that they have added processing of different data types.

https://github.com/torvalds/linux/blob/b3298500b23f0b53a8d81e0d5ad98a29db71f4f0/net/ipv4/ip_sockglue.c#L923-L932

I made a new change according to this part, please review it again, thank you!

Reply:
Looks like you dropped the range check which is important here. If someone sends and int >255 that is invalid and should error instead of truncate.

0 is also invalid unless loopback is enabled and then it is only sent on the loopback. We do not support multicast_loop so it is always invalid.